docs/admin/etcd.md

```diff
@@ -1,49 +1,163 @@
+if the apiserver will be down for the period of rollback, all of node components
+should basically restart their watches and start from “now” when apiserver is
+back. And it will be back with new resource version. That would mean that
+restarting node components is not needed.  But the assumptions here are pretty
+volatile…_
 
-## Configuration: high-level goals
+## Design
 
-Access Control: give *only* kube-apiserver read/write access to etcd. You do not
-want apiserver's etcd exposed to every node in your cluster (or worse, to the
-internet at large), because access to etcd is equivalent to root in your
-cluster.
+This section describes how we are going to do the migration given all the
+limitations we have. Note that since the code changes in Kubernetes code needed
+to support etcd v3 API are pretty local and not very sophisticated, we will not
+focus on it at all. We are only focusing on the upgrade/rollback here.
 
-Data Reliability: for reasonable safety, either etcd needs to be run as a
-[cluster](/docs/admin/high-availability/#clustering-etcd) (multiple machines each running
-etcd) or etcd's data directory should be located on durable storage (e.g., GCE's
-persistent disk). In either case, if high availability is required--as it might
-be in a production cluster--the data directory ought to be [backed up
-periodically](https://coreos.com/etcd/docs/latest/op-guide/recovery.html),
-to reduce downtime in case of corruption.
+### New etcd docker image
+We decided to completely change the way how etcd image looks like (and how it
+works). So far the docker image for etcd in version X contained only etcd and
+etcdctl binaries in version X.
 
-## Default configuration
+Going forward, the docker image for etcd in version X will contain multiple
+versions of etcd. For example 3.0.14 image will contain all: 2.2.1, 2.3.7 and
+3.0.14 binaries of etcd and etcdctl. This will allow running etcd in multiple
+different versions using exactly the same docker image.
 
-The default setup scripts use kubelet's file-based static pods feature to run etcd in a
-[pod](http://releases.k8s.io/{{page.githubbranch}}/cluster/saltbase/salt/etcd/etcd.manifest). This manifest should only
-be run on master VMs. The default location that kubelet scans for manifests is
-`/etc/kubernetes/manifests/`.
+Additionally, the image will contain our custom script for doing migration
+between versions (written by Kubernetes team) and the custom rollback tool
+(provided by CoreOS folks).
 
-## Kubernetes's usage of etcd
+### Migration script
+The migration script that will now be part of etcd docker image is a bash
+script, working as following:
 
-By default, Kubernetes objects are stored under the `/registry` key in etcd.
-This path can be prefixed by using the [kube-apiserver](/docs/admin/kube-apiserver) flag
-`--etcd-prefix="/foo"`.
+1. Detect which version of etcd we were previously running.
+   For that purpose, we have added a dedicated (metadata) file that will be
```

Comment:
Done.  I also tidied up the tense there to be more consistent, but I think it's still all over the map across the doc.